### ISSUE TITLE
create volume path validations

### ISSUE THREAD

**Issue by cpuguy83** [docker/docker#11646]
Right now volume paths are not checked until after the container is created, which can cause a container to fail to start, but not fail to be created in certain circumstances.

Need a set of validations on the volume paths prior to volume initialization.
This would include, but not limited to:

1) Is the path an empty string?
2) Is the path relative?
3) Is the path a file
4) Is the bind-path relative
...

**Comment by ctate** [comment]
Can you provide examples?

Here are the results on my end: (tested with docker 1.5)

empty string - shows error

```sh
docker create -v "" busybox
invalid value "" for flag -v:  is not an absolute path
```

relative path - shows error

```sh
docker create -v . busybox
invalid value "." for flag -v: . is not an absolute path
```

file - doesn't show error but does successfully bind to /test.txt

```sh
docker create -v /test.txt busybox
012c0d8fac59d1fd1bfe7d0d588cd6ded845320d9afee95dee0655b82a75b588
```

bind-path relative (host) - shows error

```sh
docker create -v .:/dest busybox
FATA[0000] Error response from daemon: cannot bind mount volume: . volume paths must be absolute.
docker create -v /src:. busybox
invalid value "/src:." for flag -v: . is not an absolute path
```

**Comment by cpuguy83** [comment]
Yes, thanks for looking at this, sorry this seems a bit ambiguous.
In some cases there are client side validations, so api requests aren't validated, builder, etc.
These should go into a common place.

**Comment by ctate** [comment]
Unable to replicate. Can you please provide an example in which run -v works and create -v doesn't?

**Comment by calavera** [comment]
@ctate I think what @cpuguy83 means is that those validations happen in the client side of docker, but they should probably be in the server too. The CLI checks whether the paths are correct when it calls `runconfig.Parse`: https://github.com/docker/docker/blob/455a272aef4e109ce57c20f9ea224e4d4b32ece7/runconfig/parse.go#L31

Since those validations are not in the server, everybody that wants to use the remote api needs to create those validations for their use case. Moving them to the server would make the expectations for every client consistent to the default CLI.

**Comment by resouer** [comment]
#dibs I'm writing a PR to solve this.

**Comment by rrarunan** [comment]
@resouer when you do a PR, can you please add `fixes` or `closes` keyword to it? Thanks!

**Comment by ctate** [comment]
@rrarunan will do!

**Comment by resouer** [comment]
@cpuguy83 In our PR we found that VOLUME in dockerfile is allowed to pass a relative path to dockerd (which lead ci failed), I think we should not allow that too. What do you think?

**Comment by bzz** [comment]
Just a user feedback: guys, it would be really helpful to fix this one - have just spent 2h stairing on "Unable to find image" after I have build on and it's on the list of images.. because of the malformed -v. And it also took a while to find this issue too.

**Comment by calavera** [comment]
@bzz do you mind to send us the docker version you're using and the command line that caused the problem? That would help us narrow the error cases down.

**Comment by bzz** [comment]
@calavera sure. Docker is awesome piece of software, so thank you for your time spent on making it even better!
```
Client version: 1.6.0
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 4749651
OS/Arch (client): darwin/amd64
Server version: 1.6.0
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 4749651
OS/Arch (server): linux/amd64
```
and the problem was actually very simple (and quite common, I assume) - an un-expanded env vars in the `-v` flag
```
$docker images
REPOSITORY                                TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
debian                                    jessie              bf84c1d84a8f        8 days ago          125.2 MB

$echo $OUT

$docker run -it -v "$OUT:/root/out" debian:jessie bash
Unable to find image 'debian:jessie' locally
```

The issue here is a misleading error message.

**Comment by calavera** [comment]
@bzz I think your issue might be solved in 1.7, but it could need some improvements:

```
sudo docker run -it -v ":/root/out" alpine ash
Error response from daemon: Cannot start container b7f8148ab813a2682ce2bc357f41e5008d4c27aa5aa941ec89bafcd21b9cc333: Unable to setup mount point, neither source nor volume defined
```

**Comment by calavera** [comment]
sorry, didn't want to close this issue. Daemon validations are still useful.
